library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,285 @@
+/// Encyclopedia."
+#[must_use = "this returns the result of the operation, \
+              without modifying the original"]
+#[inline(always)]
+pub const fn usize(n: usize) -> usize {
+    #[cfg(target_pointer_width = "16")]
+    {
+        u16(n as u16) as usize
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    {
+        u32(n as u32) as usize
+    }
+
+    #[cfg(target_pointer_width = "64")]
+    {
+        u64(n as u64) as usize
+    }
+}
+
+/// Generates an `i*` function that returns the [integer square root][1] of any
+/// **nonnegative** input of a specific signed integer type.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+macro_rules! signed_fn {
+    ($SignedT:ident, $UnsignedT:ident) => {
+        /// Returns the [integer square root][1] of any **nonnegative**
+        #[doc = concat!("[`", stringify!($SignedT), "`](prim@", stringify!($SignedT), ")")]
+        /// input.
+        ///
+        /// # Safety
+        ///
+        /// This results in undefined behavior when the input is negative.
+        ///
+        /// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+        /// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+        /// Encyclopedia."
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline(always)]
+        pub const unsafe fn $SignedT(n: $SignedT) -> $SignedT {
+            $UnsignedT(n as $UnsignedT) as $SignedT
+        }
+    };
+}
+
+signed_fn!(i8, u8);
+signed_fn!(i16, u16);
+signed_fn!(i32, u32);
+signed_fn!(i64, u64);
+signed_fn!(i128, u128);
+
+/// Generates a `u*` function that returns the [integer square root][1] of any
+/// input of a specific unsigned integer type.
+///
+/// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+/// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+/// Encyclopedia."
+macro_rules! unsigned_fn {
+    ($UnsignedT:ident, $HalfBitsT:ident, $stages:ident) => {
+        /// Returns the [integer square root][1] of any
+        #[doc = concat!("[`", stringify!($UnsignedT), "`](prim@", stringify!($UnsignedT), ")")]
+        /// input.
+        ///
+        /// [1]: <https://en.wikipedia.org/wiki/Integer_square_root>
+        /// "Wikipedia contributors. Integer square root. Wikipedia, The Free
+        /// Encyclopedia."
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        pub const fn $UnsignedT(mut n: $UnsignedT) -> $UnsignedT {
+            if n <= <$HalfBitsT>::MAX as $UnsignedT {
+                $HalfBitsT(n as $HalfBitsT) as $UnsignedT
+            } else {
+                const EVEN_MAKING_BITMASK: u32 = !1;
+                let normalization_shift = n.leading_zeros() & EVEN_MAKING_BITMASK;
+                n <<= normalization_shift;
```

Comment:
Sure. I've added that and will push a bit later:

<details>
<summary>The added comment</summary>

```rust
                // The normalization shift satisfies the Karatsuba square root
                // algorithm precondition "a₃ ≥ b/4" where a₃ is the most
                // significant quarter of the bits and b is the number of
                // values in `$UnsignedT`.
                //
                // In binary, b would thus be 1 followed by a count of
                // `<$UnsignedT>::BITS` 0s. b/4 would then be 1 followed by
                // `<$UnsignedT>::BITS - 2` 0s. a₃ must then be at least b/4,
                // which means that the most significant bit or its neighbor
                // must be a 1.
                //
                // The reason to shift by an even number of bits is because an
                // even number of bits produces the square root shifted to the
                // left by half of the normalization shift:
                //
                // sqrt(n << (2 * p))
                // sqrt(2.pow(2 * p) * n)
                // sqrt(2.pow(2 * p)) * sqrt(n)
                // 2.pow(p) * sqrt(n)
                // sqrt(n) << p
                //
                // Shifting by an odd number of bits leaves an ugly sqrt(2)
                // multiplied in:
                //
                // sqrt(n << (2 * p + 1))
                // sqrt(2.pow(2 * p + 1) * n)
                // sqrt(2 * 2.pow(2 * p) * n)
                // sqrt(2) * sqrt(2.pow(2 * p)) * sqrt(n)
                // sqrt(2) * 2.pow(p) * sqrt(n)
                // sqrt(2) * (sqrt(n) << p)
```
</details>